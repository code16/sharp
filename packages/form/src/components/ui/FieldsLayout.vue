<template>
    <Grid :rows="layout" v-slot="{ itemLayout:fieldLayout }">
        <template v-if="isFieldset(fieldLayout)">
            <fieldset class="SharpForm__fieldset" v-show="isFieldsetVisible(fieldLayout)">
                <div class="SharpModule__inner">
                    <div class="SharpModule__header">
                        <div class="SharpModule__title">{{fieldLayout.legend}}</div>
                    </div>
                    <div class="SharpModule__content">
                        <FieldsLayout :layout="fieldLayout.fields" v-slot="{ fieldLayout }">
                            <slot :field-layout="fieldLayout" />
                        </FieldsLayout>
                    </div>
                </div>
            </fieldset>
        </template>
        <template v-else>
            <slot :field-layout="fieldLayout" />
        </template>
    </Grid>
</template>

<script>
    import { Grid } from 'sharp-ui';

    export default {
        name: 'FieldsLayout',

        components: {
            Grid
        },

        props: {
            layout: { // 2D array fields [ligne][col]
                type: Array,
                required: true
            },
            visible : {
                type: Object,
                default: () => ({})
            }
        },

        data() {
            return {
                fieldsetMap: {}
            }
        },

        methods: {
            isFieldset(fieldLayout) {
                return !!fieldLayout.legend;
            },
            isFieldsetVisible(fieldsetLayout) {
                return (fieldsetLayout.fields || []).flat()
                    .some(fieldLayout => this.visible[fieldLayout.key]);
            }
        }
    }
</script>