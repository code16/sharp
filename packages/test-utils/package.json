{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}