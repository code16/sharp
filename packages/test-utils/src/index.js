export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}