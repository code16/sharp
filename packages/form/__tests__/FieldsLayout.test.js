import FieldsLayout from '../src/components/ui/FieldsLayout.vue';

import { shallowMount } from '@vue/test-utils';

describe('fields-layout', () => {

    const FieldMock = { name:'Field', template:'<div class="FIELD_MOCK"></div>' };
    function createWrapper({ propsData } = {}) {
        return shallowMount(FieldsLayout, {
            propsData: {
                ...propsData,
            },
            // language=Vue
            stubs: {
                Grid:
                    `<div class="GRID_MOCK">
                        <slot :item-layout="rows[0][0]" />
                    </div>`,
                FieldsLayout:
                    // render first field from given layout
                    `<div class="FIELDS_LAYOUT_MOCK">
                        <slot :field-layout="$attrs.layout[0][0]" />
                    </div>`,
                Field: FieldMock,
            },
            scopedSlots: {
                default: '<Field :data="props.fieldLayout" />',
            },
        });
    }

    test('can mount fields layout', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{ key:'title' }]
                ]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount "fieldset" fields layout', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [[
                            { key: 'title' }
                        ]]
                    }]
                ],
                visible: {
                    title: true
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount "hidden fieldset" fields layout', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [
                            [{ key: 'title' }]
                        ]
                    }]
                ],
                visible: {
                    title: false
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('expose correct props', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{ key:'list' }]
                ]
            }
        });

        expect(wrapper.find({ name: 'Field' }).vm.$attrs).toEqual({
            data: { key: 'list' }
        })
    });

    test('expose correct fieldset props', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [
                            [{ key: 'title' }]
                        ]
                    }]
                ],
            }
        });

        expect(wrapper.find({ name: 'Field' }).vm.$attrs).toEqual({
            data: { key: 'title' }
        });
    });

    test('fieldset visible', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [[{}]],
            },
        });

        wrapper.setProps({
            visible: {
                title: true,
                subtitle: false,
            },
        });

        expect(
            wrapper.vm.isFieldsetVisible({
                id: 'fieldset_1',
                legend: 'Fieldset 1',
                fields: [{ key: 'title' }, { key: 'subtitle' }],
            })
        ).toBe(true);

        wrapper.setProps({
            visible: {
                title: false,
                subtitle: false,
            },
        });

        expect(
            wrapper.vm.isFieldsetVisible({
                id: 'fieldset_1',
                legend: 'Fieldset 1',
                fields: [{ key: 'title' }, { key: 'subtitle' }],
            })
        ).toBe(false);
    });
});