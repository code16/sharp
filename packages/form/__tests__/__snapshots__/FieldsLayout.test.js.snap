// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fields-layout can mount "fieldset" fields layout 1`] = `
<div class="GRID_MOCK">
	<fieldset class="SharpForm__fieldset">
		<div class="SharpModule__inner">
			<div class="SharpModule__header">
				<div class="SharpModule__title">Fieldset</div>
			</div>
			<div class="SharpModule__content">
				<div class="FIELDS_LAYOUT_MOCK" layout="[object Object]">
					<div class="FIELD_MOCK" data="[object Object]"></div>
				</div>
			</div>
		</div>
	</fieldset>
</div>
`;

exports[`fields-layout can mount "hidden fieldset" fields layout 1`] = `
<div class="GRID_MOCK">
	<fieldset class="SharpForm__fieldset" style="display: none;">
		<div class="SharpModule__inner">
			<div class="SharpModule__header">
				<div class="SharpModule__title">Fieldset</div>
			</div>
			<div class="SharpModule__content">
				<div class="FIELDS_LAYOUT_MOCK" layout="[object Object]">
					<div class="FIELD_MOCK" data="[object Object]"></div>
				</div>
			</div>
		</div>
	</fieldset>
</div>
`;

exports[`fields-layout can mount fields layout 1`] = `
<div class="GRID_MOCK">
	<div class="FIELD_MOCK" data="[object Object]"></div>
</div>
`;
